server/lib/security/expense.ts

```diff
@@ -228,113 +228,123 @@ const getGroupedExpensesStats = (expenses: Array<Expense>) => {
-//           COUNT("id") FILTER (WHERE status = 'APPROVED') OVER (PARTITION BY "CollectiveId") AS "totalApprovedCountForCollective",
-//           (SELECT PERCENTILE_CONT(0.95) WITHIN GROUP (ORDER BY "amountInHostCurrency") FROM all_expenses WHERE "CollectiveId" = e."CollectiveId" AND status = 'PAID') AS "paidAmountP95ForCollective"
-//         FROM all_expenses e
-//       ) SELECT * FROM all_expenses_stats
-//       WHERE id IN (:expenseIds)
-//     `,
-//     {
-//       type: sequelize.QueryTypes.SELECT,
-//       replacements: { displayCurrency, collectiveIds, expenseIds },
-//     },
-//   );
-
-//   return keyBy(result, 'id');
-// };
+const getExpensesAmountsStats = async (
+  expenses: Array<Expense>,
+  displayCurrency,
+): Promise<Record<string, ExpenseAmountStats>> => {
+  const collectiveIds = uniq(expenses.map(e => e.CollectiveId));
+  const expenseIds = uniq(expenses.map(e => e.id));
+  const result: Array<ExpenseAmountStats> = await sequelize.query(
+    `
+      WITH all_expenses AS (
+        SELECT
+          e.*,
+          e.currency != :displayCurrency AS "isConvertedCurrency",
+          CASE
+            -- Simple case: expense is already in display currency
+            WHEN e.currency = :displayCurrency THEN e.amount
+            -- Convert expense to display currency
+            ELSE e.amount * COALESCE((
+                SELECT rate
+                FROM "CurrencyExchangeRates" r
+                WHERE r."from" = e.currency
+                AND r."to" = :displayCurrency
+                AND r."createdAt" <= e."createdAt"
+                ORDER BY e."createdAt" DESC -- Most recent rate that is older than the expense
+                LIMIT 1
+              ), (
+                -- Fix for old expenses where we didn't have an exchange rate stored yet: just use the oldest rate available
+                SELECT rate
+                FROM "CurrencyExchangeRates" r
+                WHERE r."from" = e.currency
+                AND r."to" = :displayCurrency
+                ORDER BY e."createdAt" ASC -- Oldest rate
+                LIMIT 1
+            ))
+            END AS "amountInHostCurrency"
+        FROM "Expenses" e
+        WHERE e."deletedAt" IS NULL
+        AND e."CollectiveId" IN (:collectiveIds)
+        AND e.status IN ('APPROVED', 'PAID')
+        ORDER BY e."createdAt" ASC
+      ), all_collective_stats AS (
+        SELECT
+          e."CollectiveId",
+          AVG("amountInHostCurrency") FILTER (WHERE status = 'PAID') AS "averageAmountForCollective",
+          SUM("amountInHostCurrency") FILTER (WHERE status = 'APPROVED') AS "totalApprovedAmountForCollective",
+          COUNT("id") FILTER (WHERE status = 'APPROVED') AS "totalApprovedCountForCollective",
+          (SELECT PERCENTILE_CONT(0.95) WITHIN GROUP (ORDER BY "amountInHostCurrency") FILTER (WHERE status = 'PAID')) AS "paidAmountP95ForCollective"
+        FROM all_expenses e
+        GROUP BY e."CollectiveId"
```

Comment:
This is the main change: rather than using windowing functions, I've moved the collective stats computation to a subquery (`WITH`) that is later joined with selected expenses. Postgres seems more efficient at grouping filters this way.